gaps/gap-17_offline_requestor_model/gap-17_offline_requestor_model.md

```diff
@@ -0,0 +1,112 @@
+---
+gap: GAP-17
+title: Offline Requestor Model
+description: Considerations and features required to implement Golem applications where Requestor is not required to be constantly connected to the network.
+author: stranger80 (@stranger80)
+status: Draft
+type: Feature
+---
+
+## Abstract
+TBC
+
+## Motivation
+The features indicated or referenced in this GAP are intended to meet following objectives:
+- Improve application resilience to network & requestor node failures
+- Enable "fire&forget", and self-sustaining application models
+
+## Specification
+Three "Offline Requestor" scenarios are considered in this GAP. 
+
+#### **A. Requestor partially connected**
+Characteristics:
+- Requestor daemon remains online until Agreement is signed and Activity started
+- Requestor daemon may then disconnect while the Activity is in progress
+- Requestor daemon may reconnect at any time, and exercise control over the Activity via ExeScript
+
+**Notes:**
+- This scenario is only possible with payment schemes which assume upfront payment or allow long intervals between payments or with self-sustained payment
+ 
+#### **B. Requestor offline ("fire&forget")** 
+Characteristics:
+- Requestor daemon remains online until Agreement is signed and Activity started
+- Requestor daemon disconnects permanently, while the Activity continues (probably until agreed computation is complete, or funds run out)
+
+**Notes:**
+- This scenario is only possible with payment schemes which assume upfront payment or allow long intervals between payments or with self-sustained payment
+ 
+#### **C. Requestor delegates control**
+Characteristics:
+- Requestor daemon remains online until Agreement is signed and Activity started
+- Requestor daemon transfers grants control rights to another node/identity, which remains online and exercises control over the Activity 
+
+**Notes:**
+- This requires **Agreement Permissions Management** feature
+- (TBD) Not sure how payments are controlled by a “delegate”...? Are all Payment actions allowed for Payments associated with a given Agreement?
+
+### Proposed features
+
+This GAP introduces following features which are aimed at enabling the "offline Reuqestor" scenarios listed above:
+- Activity Attach/Detach
+- Self-sustained Payments
+- Agreement Permissions Management
```

Comment:
Just a thought:

`Agreement Permissions Management` is a very important topic that has many use cases that have nothing to with "offline requestor". E.g. developer who creates and app that is supposed to run on Golem might prefer to rent (buy?) agreements from some third party, let's call them "broker". This way, from the developer POV:
* There is no strategy-related complexity
* There is no payment-related complexity (e.g. I broker service could just support Mastercard)
* Broker can handle some part of the risk (e.g. agreements that were ended by the provider are "free" for the end user, and broker pays-or-not-pays the provider, but from the developer POV this doesn't matter)

Actually, I wouldn't be surprised if this became the main way developers interact with Golem in the future.

-->
1. I think this topic might deserve a separate GAP ...
2. ... that should be a little more general - e.g. we might want to rent activities, not agreements?

Reply:
There is now a separate GAP for this: https://github.com/golemfactory/golem-architecture/pull/56